lrauv_ignition_plugins/src/ScienceSensorsSystem.cc

```diff
@@ -612,39 +701,304 @@ void ScienceSensorsSystemPrivate::GenerateOctrees()
+    if (octreeExceptZSlice1.nearestKSearch(_pt, 1, inds2, sqrDists2)
+      <= 0)
+    {
+      ignwarn << "Cannot find 2nd NN for trilinear interpolation" << std::endl;
+      return;
+    }
+    else
+    {
+      // Take closest point as 2nd NN
+      nnIdx2 = inds2[0];
+      minDist2 = sqrDists2[0];
+      // Get z of neighbor
+      nnZ2 = cloudExceptZSlice1.at(nnIdx2).z;
+
+      igndbg << "Slice 2 nnIdx " << nnIdx2 << ", dist " << sqrt(minDist2)
+        << std::endl;
+    }
+  }
+
+  // Step 4: Look for 4 NNs in the z slice of 2nd NN
+
+  // Get raw Eigen matrix
+  // Matrix size 3 x n
+  Eigen::MatrixXf matrixExceptZSlice1 =
+    cloudExceptZSlice1.getMatrixXfMap(3, 4, 0).block(
+      0, 0, 3, cloudExceptZSlice1.size());
+
+  // Search in z slice of 1st NN for 4 nearest neighbors in this slice
+  this->SearchInDepthSlice(_pt, nnZ2, matrixExceptZSlice1, zSlice2, zSliceInds2,
+    interpolatorInds2, interpolatorSqrDists2, _interpolators2);
+  if (interpolatorInds2.size() < 4 || interpolatorSqrDists2.size() < 4)
+  {
+    ignwarn << "Could not find enough neighbors in 2nd slice z = " << nnZ
+      << " for trilinear interpolation." << std::endl;
+  }
+}
+
+/////////////////////////////////////////////////
+void ScienceSensorsSystemPrivate::SearchInDepthSlice(
+  pcl::PointXYZ &_searchPt,
+  float _depth,
+  Eigen::MatrixXf &_points,
+  pcl::PointCloud<pcl::PointXYZ> &_zSlice,
+  std::vector<int> &_zSliceInds,
+  std::vector<int> &_interpolatorInds,
+  std::vector<float> &_interpolatorSqrDists,
+  std::vector<pcl::PointXYZ> &_interpolators,
+  int _k)
+{
+  // Find indices of points with z matching the given z. These points are on
+  // the same z slice. Logical indexing, produces Boolean array.
+  // Take only the block of the matrix containing the z values of the points.
+  Eigen::Matrix<bool, 1, Eigen::Dynamic> depthInds =
+    (_points.row(2).array() == _depth);
+
+  igndbg << "Number of points matching z == " << _depth << ": "
+         << depthInds.count() << std::endl;
+
+  // Init ret val. Pre-allocate only the number of points in the z slice
+  _zSlice = pcl::PointCloud<pcl::PointXYZ>();
+  _zSlice.points.resize(depthInds.count());
+  _zSliceInds.clear();
+  _interpolatorInds.clear();
+  _interpolatorSqrDists.clear();
+
+  int zSliceIdx = 0;
+
+  // Create a point cloud containing just the points on the z slice
+  for (int i = 0; i < depthInds.size(); ++i)
+  {
+    if (depthInds[i])
+    {
+      // Keep track of new points' indices in the original point cloud
+      _zSliceInds.push_back(i);
+
+      // Retrieve the corresponding point in the original point cloud
+      _zSlice.points[zSliceIdx++] = pcl::PointXYZ(
+        _points(0, i),
+        _points(1, i),
+        _points(2, i));
+    }
+  }
+
+  // Make sure number of points in z-slice cloud == number of pts matching z
+  assert(zSliceIdx == depthInds.count());
+
+  // Create octree for this depth slice
+  auto octree = pcl::octree::OctreePointCloudSearch<pcl::PointXYZ>(
```

Comment:
This could be pulled out too and just pass in the octree to keep the complexity of this function minimal and focused on doing the search.

Reply:
Part of 7c85554